components/segmenter/src/complex/lstm/mod.rs

```diff
@@ -144,162 +130,196 @@ impl<'l> LstmSegmenter<'l> {
-                let hunits = h_tm1.dim();
-                let embedd_dim = x_t.dim();
-                c_tm1.as_borrowed().debug_assert_dims([hunits]);
-                w.debug_assert_dims([4, hunits, embedd_dim]);
-                u.debug_assert_dims([4, hunits, hunits]);
-                b.debug_assert_dims([4, hunits]);
-            }
-
-            let mut s_t = b.to_owned();
-
-            s_t.as_mut().add_dot_3d_2(x_t, w);
-            s_t.as_mut().add_dot_3d_1(h_tm1.as_borrowed(), u);
-
-            #[allow(clippy::unwrap_used)] // first dimension is 4
-            s_t.submatrix_mut::<1>(0).unwrap().sigmoid_transform();
-            #[allow(clippy::unwrap_used)] // first dimension is 4
-            s_t.submatrix_mut::<1>(1).unwrap().sigmoid_transform();
-            #[allow(clippy::unwrap_used)] // first dimension is 4
-            s_t.submatrix_mut::<1>(2).unwrap().tanh_transform();
-            #[allow(clippy::unwrap_used)] // first dimension is 4
-            s_t.submatrix_mut::<1>(3).unwrap().sigmoid_transform();
-
-            #[allow(clippy::unwrap_used)] // first dimension is 4
-            c_tm1.convolve(
-                s_t.as_borrowed().submatrix(0).unwrap(),
-                s_t.as_borrowed().submatrix(2).unwrap(),
-                s_t.as_borrowed().submatrix(1).unwrap(),
-            );
+    /// Create an LSTM based break iterator for a UTF-16 string.
+    pub(super) fn segment_utf16(&'l self, input: &[u16]) -> impl Iterator<Item = usize> + 'l {
+        let input_seq = if let Some(grapheme) = self.grapheme {
+            GraphemeClusterSegmenter::new_and_segment_utf16(input, grapheme)
+                .collect::<Vec<usize>>()
+                .windows(2)
+                .map(|chunk| {
+                    let range = if let [first, second, ..] = chunk {
+                        *first..*second
+                    } else {
+                        unreachable!()
+                    };
+                    let grapheme_cluster = if let Some(grapheme_cluster) = input.get(range) {
+                        grapheme_cluster
+                    } else {
+                        return self.dic.len() as u16;
+                    };
 
-            #[allow(clippy::unwrap_used)] // first dimension is 4
-            h_tm1.mul_tanh(s_t.as_borrowed().submatrix(3).unwrap(), c_tm1.as_borrowed());
+                    // The maximum UTF-8 size of a grapheme cluster seems to be 41 bytes
+                    let mut i = 0;
+                    let mut buf = [0; 41];
+
+                    #[allow(clippy::unwrap_used)]
+                    // debug_asserting whether my assumption is correct
+                    decode_utf16(grapheme_cluster.iter().copied()).for_each(|c| {
+                        debug_assert!(i < 37);
+                        i += c
+                            .unwrap_or(REPLACEMENT_CHARACTER)
+                            .encode_utf8(&mut buf[i..])
+                            .len()
+                    });
```

Comment:
thoughts about this?

Reply:
_scratching head_

What is this code for and why did it not exist before?

I'll admit that I spot-checked that this code was a refactor of the old code, but I didn't see this change. Thanks for highlighting

Reply:
The u16 code path previously required reallocating the entire `&[u16]` as a `String`. I've split the `str`/`&[u16]`-to-dictionary mapping so we can work on the input without reallocating. For `str` this is the existing code path, for `u16` it's this for grapheme cluster models, and the `decode_utf16` iterator below.

Reply:
ok I see. Thanks for the extra context.

Can you avoid putting a maximum length on the grapheme cluster by using `self.dic.get_copied_by`? In the predicate function, you could make a UTF-8 iterator over both the grapheme cluster and the dictionary key and then compare them.

Reply:
Hmm, this would require recalculating UTF-8 at every step of the binary search...

Reply:
Yeah I'm not sure which is more efficient, except that iterating UTF-16 as UTF-8 _should_ be a very cheap operation.

The 41 bytes may be sufficient for now but it's unlikely to always be enough.

Maybe use a smallvec?

Reply:
There's also no function to iterate over wtf-8 (`UnvalidatedStr`) with replacement characters

Reply:
... yet https://doc.rust-lang.org/std/str/struct.Utf8Chunks.html

Reply:
Ok check this out. The code has no test coverage

Reply:
Can you cover it? Just modify the grapheme test to run both the utf8 and utf16 code path. This is more important than it was before because the code paths have diverged.

Reply:
Done. The breakpoints are from main, I haven't manually validated them.